wolfcrypt/src/sakke.c

```diff
@@ -0,0 +1,6774 @@
+ *
+ * q must not be the same object as r.
+ * Calculation:
+ *   r.x = (p.x * q.x) - (p.y * q.y)
+ *   r.y = (p.x * q.y) + (p.y * q.x)
+ * But qx = 1
+ *   r.x = p.x - (p.y * q.y)
+ *   r.y = (p.x * q.y) + p.y
+ *
+ * @param  [in]   p      MP projective integer - first value to multiply.
+ * @param  [in]   q      MP projective integer - second value to multiply.
+ * @param  [in]   prime  MP integer that is the modulus of the field.
+ * @param  [in]   mp     Multiplier to use when converting from Montgomery form.
+ * @param  [out]  r      Result of operation.
+ * @param  [in]   t1     MP integer temporary.
+ * @param  [in]   t2     MP integer temporary.
+ * @return  0 on success.
+ * @return  MEMORY_E when dynamic memory allocation fails.
+ * @return  Other -ve value on internal failure.
+ */
+static int sakke_proj_mul_qx1(mp_proj* p, mp_int* q, mp_int* prime, mp_digit mp,
+        mp_proj* r, mp_int* t1, mp_int* t2)
+{
+    int err;
+
+    /* t1 = p.x * q.y */
+    err = sakke_mont_mulmod(p->x, q, prime, t1, mp);
+    /* t2 = p.y * q.y */
+    if (err == 0) {
+        err = sakke_mont_mulmod(p->y, q, prime, t2, mp);
+    }
+    /* r.x = p.x - (p.y * q.y) */
+    if (err == 0) {
+        err = sakke_submod(p->x, t2, prime, r->x);
+    }
+    /* r.y = (p.x * q.y) + p.y */
+    if (err == 0) {
+        err = sakke_addmod(t1, p->y, prime, r->y);
+    }
+
+    return err;
+}
+
+/*
+ * Calculate the gradient of line hrough P, P and [-2]P.
+ *
+ * @param  [in]   p      ECC point - first point on the elliptic curve.
+ * @param  [in]   t2     MP integer temporary.
+ * @param  [in]   prime  MP integer that is the modulus of the field.
+ * @param  [in]   mp     Multiplier to use when converting from Montgomery form.
+ * @param  [out]  l      MP integer representing gradient.
+ * @param  [out]  z2     MP integer representing p.z^2.
+ */
+static int sakke_calc_dbl_rx(ecc_point* p, ecc_point* q, mp_int* prime,
+        mp_digit mp, mp_proj* r, mp_int* l, mp_int* z2)
+{
+    int err;
+    mp_int* t = r->y;
+
+    /* z2 = p.z^2 */
+    err = sakke_mont_sqrmod(p->z, prime, z2, mp);
+    /* t = p.x + p.z^2 */
+    if (err == 0) {
+        err = sakke_addmod(p->x, z2, prime, l);
+    }
+    /* r1 = p.x - p.z^2 */
+    if (err == 0) {
+        err = sakke_submod(p->x, z2, prime, t);
+    }
+    /* l = ((p.x - p.z^2) * (p.x + p.z^2) = p.x ^ 2 - p.z^4  */
+    if (err == 0) {
+        err = sakke_mont_mulmod(l, t, prime, l, mp);
+    }
+    /* l = ((p.x ^ 2) - p.z^4) * 3 */
+    if (err == 0) {
+        err = sakke_tplmod(l, prime, l);
+    }
+    /* t = q.x * p.z^2 */
+    if (err == 0) {
+        err = sakke_mont_mulmod(q->x, z2, prime, t, mp);
+    }
+    /* t = p.x + q.x * p.z^2 */
+    if (err == 0) {
+        err = sakke_addmod(p->x, t, prime, t);
+    }
+    /* r.x = l * (p.x + q.x * p.z^2) */
+    if (err == 0) {
+        err = sakke_mont_mulmod(l, t, prime, r->x, mp);
+    }
+
+    return err;
+}
+
+#ifdef WOLFSSL_SMALL
```

Comment:
This file has three uses of new macro `WOLFSSL_SMALL` with no test case. Should it be `WOLFSSL_SP_SMALL`?

Reply:
Changed to WOLFSSL_SAKKE_SMALL